plugins/storage/image/swift/src/main/java/org/apache/cloudstack/storage/datastore/driver/SwiftImageStoreDriverImpl.java

```diff
@@ -100,12 +102,28 @@ public String createEntityExtractUrl(DataStore store, String installPath, ImageF
 
     @Override
     public void createAsync(DataStore dataStore, DataObject data, AsyncCompletionCallback<CreateCmdResult> callback) {
-        Long maxTemplateSizeInBytes = getMaxTemplateSizeInBytes();
-        VirtualMachineTemplate tmpl = _templateDao.findById(data.getId());
+
+        DownloadCommand downloadCommand = null;
+        if (data.getType() == DataObjectType.TEMPLATE) {
+            Long maxTemplateSizeInBytes = getMaxTemplateSizeInBytes();
+            downloadCommand = new DownloadCommand((TemplateObjectTO) (data.getTO()), maxTemplateSizeInBytes);
+        }else if (data.getType() == DataObjectType.VOLUME){
+            Long maxDownloadSizeInBytes = getMaxVolumeSizeInBytes();
+            VolumeInfo volumeInfo = (VolumeInfo) data;
+            RegisterVolumePayload payload = (RegisterVolumePayload) volumeInfo.getpayload();
+            ImageFormat format = ImageFormat.valueOf(payload.getFormat());
+            downloadCommand = new DownloadCommand((VolumeObjectTO) (data.getTO()), maxDownloadSizeInBytes, payload.getChecksum(), payload.getUrl(), format);
+        }
+
+        if (downloadCommand == null){
+            String errMsg = "Unable to build download command, DataObject is of neither VOLUME or TEMPLATE type";
```

Comment:
Suggestion:

Improve logging by adding DataObject data in the message.

Reply:
thanks will do it.